pkg/operator/encryption/shared_encryption.go

```diff
@@ -145,19 +151,54 @@ type keysState struct {
 	secretsMigratedNo  []*corev1.Secret
 
 	lastMigrated      *corev1.Secret
-	lastMigratedKey   apiserverconfigv1.Key
+	lastMigratedKey   keyAndMode
 	lastMigratedKeyID uint64
 }
 
 // TODO docs
-// TODO expand fields to include key mode
 type groupResourceKeys struct {
-	writeKey apiserverconfigv1.Key
-	readKeys []apiserverconfigv1.Key
+	writeKey keyAndMode
+	readKeys []keyAndMode
 }
 
 func (k groupResourceKeys) hasWriteKey() bool {
-	return len(k.writeKey.Name) > 0 && len(k.writeKey.Secret) > 0
+	return len(k.writeKey.key.Name) > 0 && len(k.writeKey.key.Secret) > 0
+}
+
+type keyAndMode struct {
+	key  apiserverconfigv1.Key
+	mode mode
+}
+
+// mode is the value associated with the encryptionSecretMode annotation
+type mode string
+
+// The current set of modes that are supported along with the default mode that is used.
+// These values are encoded into the secret and thus must not be changed.
+// Strings are used over iota because they are easier for a human to understand.
+const (
+	aescbc    mode = "aescbc"    // available from the first release, see defaultMode below
+	secretbox mode = "secretbox" // available from the first release, see defaultMode below
+
+	// Changing this value requires caution to not break downgrades.
+	// Specifically, if some new mode is released in version X, that new mode cannot
+	// be used as the defaultMode until version X+1.  Thus on a downgrade the operator
+	// from version X will still be able to honor the observed encryption state
+	// (and it will do a key rotation to force the use of the old defaultMode).
+	defaultMode = aescbc
+)
+
+var modeToNewKeyFunc = map[mode]func() []byte{
+	aescbc:    newAES256Key,
+	secretbox: newAES256Key, // secretbox requires a 32 byte key so we can reuse the same function here
+}
+
+func newAES256Key() []byte {
+	b := make([]byte, 32) // AES-256 == 32 byte key
+	if _, err := rand.Read(b); err != nil {
+		panic(err) // rand should never fail
```

Comment:
Trace it all the way:

1. rand Reader on unix https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/crypto/rand/rand_unix.go#L34
1. defers to https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/crypto/rand/rand_unix.go#L57
1. getrandom set https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/crypto/rand/rand_batched.go#L15
1. it gets data directly since we ask for 32 bytes only https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/crypto/rand/rand_batched.go#L28
1. and call getrandom https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/crypto/rand/rand_batched.go#L36-L40
1. with no flags set https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/internal/syscall/unix/getrandom_linux.go#L28
1. and length is based on buffer https://github.com/golang/go/blob/28f8f55bd29f997fc814782f161ae09520955a8d/src/internal/syscall/unix/getrandom_linux.go#L35-L38

Meaning none of the errors can occur:

1. we block forever so no - `EAGAIN` The requested entropy was not available, and getrandom() would have blocked if the GRND_NONBLOCK flag was not set.
1. this is Go so no - `EFAULT` The address referred to by buf is outside the accessible address space.
1. cannot happen because we ask for a small enough buffer size (see below) - `EINTR` The call was interrupted by a signal handler
1. flag is always 0 so no - `EINVAL` An invalid flag was specified in flags.
1. modern kernel so no - `ENOSYS` The glibc wrapper function for getrandom() determined that the underlying kernel does not implement this system call.

> When reading from the urandom source (GRND_RANDOM is not set), getrandom() will block until the entropy pool has been initialized (unless the GRND_NONBLOCK flag was specified).
> ...
> If the entropy pool has been initialized and the request size is small (buflen <= 256), then getrandom() will not fail with EINTR. Instead, it will return all of the bytes that have been requested.

I do not feel any particular need to guard against `entropy pool has been initialized` since even in CI where such a case is more likely, I have never seen it happen in all of 3.x (and such a panic would have triggered a CI warning via the sentry integration we had back then).

I really prefer to treat never-before-seen failures in crypto related code to be catastrophic via a `panic`.

Reply:
wow, give me some time to analyze it but thanks I really appreciate the effort :)